source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,542 @@
+
+  // Close and delete any other connections.
+  auto it = connections_.begin();
+  while (it != connections_.end()) {
+    if (it->get() != &(wrapper->connection())) {
+      (*it)->removeConnectionCallbacks(*wrapper);
+      (*it)->close(ConnectionCloseType::NoFlush);
+      it = connections_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+  ASSERT(connections_.size() == 1);
+  callbacks_wrappers_.clear();
+
+  // Apply post-connect state to the final socket.
+  for (auto cb : post_connect_state_.connection_callbacks_) {
+    if (cb) {
+      connections_[0]->addConnectionCallbacks(*cb);
+    }
+  }
+
+  for (const auto& cb : post_connect_state_.bytes_sent_callbacks_) {
+    connections_[0]->addBytesSentCallback(cb);
+  }
+
+  if (event == ConnectionEvent::Connected) {
+    // Apply post-connect state which is only connections which have succeeded.
+    for (auto& filter : post_connect_state_.filters_) {
+      connections_[0]->addFilter(filter);
+    }
+    for (auto& filter : post_connect_state_.write_filters_) {
+      connections_[0]->addWriteFilter(filter);
+    }
+    for (auto& filter : post_connect_state_.read_filters_) {
+      connections_[0]->addReadFilter(filter);
+    }
+    if (post_connect_state_.initialize_read_filters_.has_value() &&
+        post_connect_state_.initialize_read_filters_.value()) {
+      bool initialized = connections_[0]->initializeReadFilters();
+      ASSERT(initialized);
+    }
+    if (post_connect_state_.read_disable_count_.has_value()) {
+      for (int i = 0; i < post_connect_state_.read_disable_count_.value(); ++i) {
+        connections_[0]->readDisable(true);
+      }
+    }
+
+    if (post_connect_state_.write_buffer_.has_value()) {
+      // write_buffer_ and end_stream_ are both set together in write().
+      ASSERT(post_connect_state_.end_stream_.has_value());
+      connections_[0]->write(*post_connect_state_.write_buffer_.value(),
+                             post_connect_state_.end_stream_.value());
+    }
+  }
+}
+
+void HappyEyeballsConnectionImpl::cleanupWrapperAndConnection(ConnectionCallbacksWrapper* wrapper) {
+  for (auto it = connections_.begin(); it != connections_.end();) {
+    if (it->get() == &(wrapper->connection())) {
+      (*it)->close(ConnectionCloseType::NoFlush);
+      it = connections_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+
+  for (auto it = callbacks_wrappers_.begin(); it != callbacks_wrappers_.end();) {
+    if (it->get() == wrapper) {
+      it = callbacks_wrappers_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+}
+
+void HappyEyeballsConnectionImpl::onAboveWriteBufferHighWatermark(
+    ConnectionCallbacksWrapper* /*wrapper*/) {
+  NOT_REACHED_GCOVR_EXCL_LINE;
+}
+
+void HappyEyeballsConnectionImpl::onBelowWriteBufferLowWatermark(
+    ConnectionCallbacksWrapper* /*wrapper*/) {
+  NOT_REACHED_GCOVR_EXCL_LINE;
+}
+
+void HappyEyeballsConnectionImpl::onWriteBufferHighWatermark() { ASSERT(!connect_finished_); }
```

Comment:
hm, if we're not going to pass this up to the writer we should comment why.

Reply:
Oh, interesting. In the previous version of this PR, I was saving this signal to a local `above_high_watermark_` member. I removed that now that we're delegating to the write buffer to implement `aboveHighWatermark()`. But that reveals that we're not doing anything with this signal, and that's likely problematic. Instead we should be invoking the connection callback in this case. Good catch! Fixed and tested.